lib/ansible/parsing/vault/__init__.py

```diff
@@ -1078,8 +1078,29 @@ def write_data(self, data, filename, shred=True):
                     self._shred_file(filename)
                 else:
                     os.remove(filename)
-            with open(filename, "wb") as fh:
-                fh.write(b_file_data)
+
+            # Ensure we create the file and it is opened with specified mode
+            #
+            # Set umask to ensure the file will be created with the correct permissions
+            current_umask = os.umask(0o077)
+
+            # Create a file with default permissions
+            try:
+                fd = os.open(filename, os.O_CREAT | os.O_EXCL | os.O_RDWR, mode)
+            except FileExistsError:
+                os.unlink(filename)
+                fd = os.open(filename, os.O_CREAT | os.O_EXCL | os.O_RDWR, mode)
+
+            try:
+                os.write(fd, b_file_data)
+            except Exception:
```

Comment:
Small nitpick - can we implement `finally` here ?

Reply:
That's a great idea. I needed to do some more reading on the subtleties of exception handling to make sure the handler _always_ gets closed.

Reply:
I think I have this done much better now using a `finally` statement. I also added some better error handling and messaging.